RustBridge/RustBridgeImplementation.cpp

```diff
@@ -0,0 +1,385 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2022 Metrological
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "RustBridge.h"
+
+namespace WPEFramework {
+
+	namespace Plugin {
+
+		namespace Rust {
+			struct RequestContext {
+				uint32_t    id;
+				const char* auth_token;
+			};
+
+			struct PluginContext;
+
+			struct Plugin;
+		}
+	}
+}
+
+extern "C" void wpe_send_to(uint32_t id, const char* json, WPEFramework::Plugin::Rust::PluginContext* p_ctx);
+
+namespace WPEFramework {
+
+	namespace Plugin {
+
+		class RustBridgeImplementation :  public Exchange::IRustBridge {
+		private:
+			class Config : public Core::JSON::Container {
+			public:
+				Config(const Config&) = delete;
+				Config& operator=(const Config&) = delete;
+				Config() 
+					: Core::JSON::Container()
+					, ModuleName()
+				{
+					Add(_T("module"), &ModuleName);
+				}
+				~Config() override = default;
+				
+			public:
+				Core::JSON::String ModuleName;
+			};
+			
+			class Connector {
+			private:
+				using fn_SendTo = void (*)(uint32_t, const char *, Rust::PluginContext *);
+				using fn_Create = Rust::Plugin *(*)(const char *name, fn_SendTo send_to, Rust::PluginContext *, void *);
+				using fn_Destroy = void (*)(Rust::Plugin *p);
+				using fn_Init = void (*)(Rust::Plugin *p, const char *json);
+				using fn_Invoke = void (*)(Rust::Plugin *p, const char *json_req, Rust::RequestContext req_ctx);
+
+				// TODO: Discuss this since I think these are not needed. I assume it is to send events using the 
```

Comment:
Taking this Metrological plugin into the RDKServices as a replacement for the RustAdapter is oke, but when we did the implementation we would like to propose some fundamental changes to improve the performance. One of them is to define a new method that reports to the bridge the exposed methods. This would handle methods that have no meaning/counter part in rust already in this plugin (avoiding needless process hops).
Might be good, as we make this now, I assume, production worthy to pick up these discussions again.